In [1]:
import torch, time
device = torch.device("cuda"); print(device.type)
from holo.calc import product
from holo.profilers import SimpleProfiler, Profiler, prettyTime
from holo.prettyFormats import prettyDataSizeOctes, get_prettyTime_Formater
import calflops
import attrs

cuda


In [240]:
def resetMemStats():
    torch.cuda.empty_cache()
    torch.cuda.reset_peak_memory_stats()
    torch.cuda.reset_accumulated_memory_stats()
    time.sleep(0.1)

def memDiag(name:str):
    print(f"memory status of {name}")
    mem = torch.cuda.max_memory_allocated()
    print(f"torch.cuda.max_memory_allocated: {prettyDataSizeOctes(mem)}")
    print(f"torch.cuda.max_memory_reserved: {prettyDataSizeOctes(torch.cuda.max_memory_reserved())}")
    print("---"*5)
    resetMemStats()
    return f"{int(mem/1e6):_d}Mo"

def showTimeStat(
        sp:SimpleProfiler, nbIters:int, baseBatchSize:int, convShape:tuple[int, int], 
        nbExperts:int, nbFeatures:int, memUsage:str):
    batchSize = baseBatchSize * convShape[0] * convShape[1]
    tPerIter = (sp.time() / nbIters)
    tPerElts = (tPerIter / batchSize)
    tPerEltsExperts = (tPerElts / nbExperts)
    ptNs = get_prettyTime_Formater("ns")
    ptMs = get_prettyTime_Formater("ms")
    print(f"{sp.name} -> {prettyTime(tPerIter)} | {ptNs(tPerElts)}/elts | {ptNs(tPerEltsExperts)}/(elts*experts)")
    pt2Ns = lambda t: f"{t*1e9:.0f}ns"
    print(f"| {baseBatchSize}\\*{convShape[0]}*{convShape[1]} | {nbExperts} | {nbFeatures} "
          f"| {ptMs(tPerIter)} | {pt2Ns(tPerElts)} | {pt2Ns(tPerEltsExperts)} | {memUsage} |")

In [305]:
device = torch.device("cuda"); print(device.type)

def cell(baseBatchSize:int, convShape:tuple[int, int], nbExperts:int, nbFeatures:int, nbOut:int):
    print(f"*** unsing: {baseBatchSize=}, {convShape=}, {nbExperts=}, {nbFeatures=}, {nbOut=} ****")
    resetMemStats()
    batchSize = baseBatchSize * convShape[0] * convShape[1]
    nbHidden = max(nbFeatures, nbOut) * 4
    _profbefore = Profiler(["topK", "softMax", "expertsWeigths"])
    experts = [
        torch.nn.Sequential(
            torch.nn.Linear(nbFeatures, nbHidden),
            torch.nn.ReLU(), torch.nn.Linear(nbHidden, nbOut)).to(device)
        for _ in range(nbExperts)]
    nbParams = sum(sum(product(p.shape) for p in e.parameters()) for e in experts)
    print(f"nb params: {nbParams:_d} (nbHidden: {nbHidden:_d})")
    k = 2
    
    gatingLogits = torch.randn((batchSize, nbExperts), device=device) * 2 # (batchSize, nbExperts)
    with _profbefore.mesure("topK"):
        gatingLogits_k, topK_indices = torch.topk(gatingLogits, k=k, dim=1) # (batchSize, k), (batchSize, k)
    with _profbefore.mesure("softMax"):
        gatingProb = torch.softmax(gatingLogits_k, dim=1) # (batchSize, k)
    with _profbefore.mesure("expertsWeigths"):
        allExperts_weigths = torch.zeros((batchSize, nbExperts), device=device) # (batchSize, nbExperts)
        allExperts_weigths.scatter_(dim=1, index=topK_indices, src=gatingProb)
    
    totbefore = sum(_profbefore.totalTimes().values())
    for c in _profbefore.categories:
        print(f"{c}:{_profbefore.totalMesure(c) / totbefore:.2%}", end=", ")
    print(_profbefore.pretty_totalTimes())
    
    datasIn = torch.randn((batchSize, nbFeatures), device=device)

    _prof = Profiler(["A", "B", "C1", "C2", "D", "E", "F"])
    memDiag("initial datas")

    N = 10
    with SimpleProfiler("efficient") as sp1:#, torch.no_grad():
        for i in range(N*0):
            datasOut = torch.zeros((batchSize, nbOut), device=device)
            for iExpert in range(nbExperts):
                with _prof.mesure("A"):
                    where = (topK_indices == iExpert).any(dim=1)
                with _prof.mesure("B"):
                    expertUse, *_ = torch.where(where)
                if expertUse.shape[0] == 0:
                    continue
                with _prof.mesure("C1"):
                    inin = datasIn[expertUse]
                with _prof.mesure("C2"):
                    expertOuts = experts[iExpert](inin)
                with _prof.mesure("D"):
                    expertGates = allExperts_weigths[expertUse, iExpert, None]
                with _prof.mesure("E"):
                    res = expertGates * expertOuts
                with _prof.mesure("F"):
                    datasOut.index_add_(dim=0, index=expertUse, source=res)
                del where, expertUse, _, inin, expertGates, res
            del datasOut
    def block():
        for c in _prof.categories:
            print(f"{c}:{_prof.totalMesure(c) / sp1.time():.2%}", end=", ")
        print(_prof.pretty_avgTimes())
        tt = sp1.time(); tts = f"{tt/N*1e3:.3f}ms"
        tB = _prof.totalMesure("B"); tBs = f"{tB*1e3/N/nbExperts:.3f}ms"
        tC2 = _prof.totalMesure("C2"); tC2s = f"{tC2*1e3/N/nbExperts:.3f}ms"
        twBC2 = (tt-tB-tC2); twBC2s = f"{twBC2*1e3/N/nbExperts:.3f}ms"
        ttas = f"{tt/nbExperts/N*1e3:.3f}ms"
        print(f"| {baseBatchSize}\\*{convShape[0]}*{convShape[1]} | {nbExperts} | {nbFeatures} "
          f"| {nbParams:_d} | {tts} | {ttas} | {tBs} | {tC2s} | {twBC2s} |")
    #block(); del block
    mem1 = memDiag("after effiecient topK")
    showTimeStat(sp1, nbIters=N, baseBatchSize=baseBatchSize, convShape=convShape, nbExperts=nbExperts, nbFeatures=nbFeatures, memUsage=mem1)

    with SimpleProfiler("computeAll") as sp2:#, torch.no_grad():
        for i in range(N):
            expertsOut = torch.stack([expert(datasIn) for expert in experts], dim=2)
            mergedOuts = torch.sum(allExperts_weigths[:, None, :] * expertsOut, dim=2)
            del expertsOut, mergedOuts
    mem2 = memDiag("after compute all")
    showTimeStat(sp2, nbIters=N, baseBatchSize=baseBatchSize, convShape=convShape, nbExperts=nbExperts, nbFeatures=nbFeatures, memUsage=mem2)
    
    with SimpleProfiler("computeAll V2") as sp3:#, torch.no_grad():
        for i in range(N):
            datasOut = torch.zeros((batchSize, nbOut), device=device)
            for iExpert in range(nbExperts):
                expertOut = experts[iExpert](datasIn)
                datasOut += (allExperts_weigths[:, iExpert, None] * expertOut)
                del expertOut
            del datasOut
    mem3 = memDiag("after compute all V2")
    showTimeStat(sp3, nbIters=N, baseBatchSize=baseBatchSize, convShape=convShape, nbExperts=nbExperts, nbFeatures=nbFeatures, memUsage=mem3)

cell(baseBatchSize=256, convShape=(7, 7), nbExperts=128, nbFeatures=128, nbOut=128); del cell


cuda
*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=128, nbFeatures=128, nbOut=128 ****
nb params: 16_859_136 (nbHidden: 512)
topK:44.93%, softMax:15.37%, expertsWeigths:39.70%, {'topK': '53.200 μs', 'softMax': '18.200 μs', 'expertsWeigths': '47.000 μs'}
memory status of initial datas
torch.cuda.max_memory_allocated: 106.212 Mo
torch.cuda.max_memory_reserved: 115.343 Mo
---------------
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 106.212 Mo
torch.cuda.max_memory_reserved: 115.343 Mo
---------------
efficient -> 0.360 μs | 0.029 ns/elts | 0.0 ns/(elts*experts)
| 256\*7*7 | 128 | 128 | 0.0 ms | 0ns | 0ns | 106Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 5.045 Go
torch.cuda.max_memory_reserved: 6.149 Go
---------------
computeAll -> 19.267 ms | 1535.963 ns/elts | 12.0 ns/(elts*experts)
| 256\*7*7 | 128 | 128 | 19.267 ms | 1536ns | 12ns | 5_045Mo |
memory status of after compute all V2
torch.cuda.max_memory_allocated: 3.427

```raw
*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=6, nbFeatures=64, nbOut=64 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 13.53 Mo
torch.cuda.max_memory_reserved: 23.069 Mo
---------------
A:4.90%, B:51.64%, C1:3.82%, C2:22.51%, D:5.25%, E:2.60%, F:2.80%, {'A': '3.638 ms', 'B': '38.351 ms', 'C1': '2.834 ms', 'C2': '16.714 ms', 'D': '3.896 ms', 'E': '1.930 ms', 'F': '2.082 ms'}
efficient -> 3.713 ms | 296.013 ns/elts | 49.335 ns/(elts*experts)
| 256*7*7 | 6 | 64 | 296ns | 49ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 64.473 Mo
torch.cuda.max_memory_reserved: 90.178 Mo
---------------
computeAll -> 0.855 ms | 68.193 ns/elts | 11.366 ns/(elts*experts)
| 256*7*7 | 6 | 64 | 68ns | 11ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 138.886 Mo
torch.cuda.max_memory_reserved: 205.521 Mo
---------------

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=24, nbFeatures=64, nbOut=64 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 18.475 Mo
torch.cuda.max_memory_reserved: 25.166 Mo
---------------
A:6.34%, B:37.91%, C1:5.06%, C2:29.05%, D:6.71%, E:3.38%, F:3.63%, {'A': '14.268 ms', 'B': '85.378 ms', 'C1': '11.405 ms', 'C2': '65.434 ms', 'D': '15.108 ms', 'E': '7.620 ms', 'F': '8.172 ms'}
efficient -> 11.261 ms | 897.71 ns/elts | 37.405 ns/(elts*experts)
| 256*7*7 | 24 | 64 | 898ns | 37ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 62.982 Mo
torch.cuda.max_memory_reserved: 88.08 Mo
---------------
computeAll -> 3.107 ms | 247.685 ns/elts | 10.32 ns/(elts*experts)
| 256*7*7 | 24 | 64 | 248ns | 10ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 486.748 Mo
torch.cuda.max_memory_reserved: 645.923 Mo
---------------

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=64, nbFeatures=64, nbOut=64 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 27.904 Mo
torch.cuda.max_memory_reserved: 31.457 Mo
---------------
A:6.02%, B:40.98%, C1:4.77%, C2:27.67%, D:6.31%, E:3.23%, F:3.39%, {'A': '41.205 ms', 'B': '280.654 ms', 'C1': '32.650 ms', 'C2': '189.491 ms', 'D': '43.217 ms', 'E': '22.141 ms', 'F': '23.240 ms'}
efficient -> 34.244 ms | 2729.95 ns/elts | 42.655 ns/(elts*experts)
| 256*7*7 | 64 | 64 | 2730ns | 43ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 71.166 Mo
torch.cuda.max_memory_reserved: 94.372 Mo
---------------
computeAll -> 10.371 ms | 826.763 ns/elts | 12.918 ns/(elts*experts)
| 256*7*7 | 64 | 64 | 827ns | 13ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 1.265 Go
torch.cuda.max_memory_reserved: 1.636 Go
---------------

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=128, nbFeatures=64, nbOut=64 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 42.813 Mo
torch.cuda.max_memory_reserved: 60.817 Mo
---------------
A:6.07%, B:41.92%, C1:4.95%, C2:26.27%, D:6.40%, E:3.27%, F:3.43%, {'A': '85.980 ms', 'B': '593.813 ms', 'C1': '70.071 ms', 'C2': '372.172 ms', 'D': '90.628 ms', 'E': '46.317 ms', 'F': '48.653 ms'}
efficient -> 70.829 ms | 5646.42 ns/elts | 44.113 ns/(elts*experts)
| 256*7*7 | 128 | 64 | 5646ns | 44ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 86.07 Mo
torch.cuda.max_memory_reserved: 102.76 Mo
---------------
computeAll -> 30.070 ms | 2397.125 ns/elts | 18.728 ns/(elts*experts)
| 256*7*7 | 128 | 64 | 2397ns | 19ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 2.513 Go
torch.cuda.max_memory_reserved: 3.236 Go
---------------

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=256, nbFeatures=64, nbOut=64 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 71.77 Mo
torch.cuda.max_memory_reserved: 85.983 Mo
---------------
A:6.56%, B:38.90%, C1:5.21%, C2:27.07%, D:6.89%, E:3.53%, F:3.74%, {'A': '155.690 ms', 'B': '923.658 ms', 'C1': '123.689 ms', 'C2': '642.800 ms', 'D': '163.606 ms', 'E': '83.845 ms', 'F': '88.772 ms'}
efficient -> 118.720 ms | 9464.308 ns/elts | 36.97 ns/(elts*experts)
| 256*7*7 | 256 | 64 | 9464ns | 37ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 115.053 Mo
torch.cuda.max_memory_reserved: 127.926 Mo
---------------
computeAll -> 91.033 ms | 7257.062 ns/elts | 28.348 ns/(elts*experts)
| 256*7*7 | 256 | 64 | 7257ns | 28ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 5.008 Go
torch.cuda.max_memory_reserved: 6.445 Go
---------------

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=512, nbFeatures=64, nbOut=64 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 131.408 Mo
torch.cuda.max_memory_reserved: 144.703 Mo
---------------
A:6.05%, B:43.11%, C1:4.78%, C2:25.23%, D:6.27%, E:3.27%, F:3.44%, {'A': '312.127 ms', 'B': '2.224 sec', 'C1': '246.804 ms', 'C2': '1.302 sec', 'D': '323.344 ms', 'E': '168.561 ms', 'F': '177.604 ms'}
efficient -> 257.920 ms | 20561.24 ns/elts | 40.159 ns/(elts*experts)
| 256*7*7 | 512 | 64 | 20561ns | 40ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 174.886 Mo
torch.cuda.max_memory_reserved: 186.647 Mo
---------------
computeAll -> 217.858 ms | 17367.543 ns/elts | 33.921 ns/(elts*experts)
| 256*7*7 | 512 | 64 | 17368ns | 34ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 10.0 Go
torch.cuda.max_memory_reserved: 12.847 Go
---------------




######################################################################################



*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=6, nbFeatures=128, nbOut=128 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 19.646 Mo
torch.cuda.max_memory_reserved: 25.166 Mo
---------------
A:4.79%, B:51.24%, C1:4.05%, C2:22.61%, D:5.06%, E:2.68%, F:2.75%, {'A': '3.997 ms', 'B': '42.756 ms', 'C1': '3.382 ms', 'C2': '18.869 ms', 'D': '4.220 ms', 'E': '2.238 ms', 'F': '2.298 ms'}
efficient -> 4.172 ms | 332.598 ns/elts | 55.433 ns/(elts*experts)
| 256*7*7 | 6 | 128 | 333ns | 55ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 112.067 Mo
torch.cuda.max_memory_reserved: 132.121 Mo
---------------
computeAll -> 1.099 ms | 87.591 ns/elts | 14.598 ns/(elts*experts)
| 256*7*7 | 6 | 128 | 88ns | 15ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 270.259 Mo
torch.cuda.max_memory_reserved: 367.002 Mo
---------------

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=24, nbFeatures=128, nbOut=128 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 30.936 Mo
torch.cuda.max_memory_reserved: 35.652 Mo
---------------
A:5.98%, B:40.11%, C1:4.87%, C2:28.10%, D:6.33%, E:3.30%, F:3.42%, {'A': '16.464 ms', 'B': '110.421 ms', 'C1': '13.397 ms', 'C2': '77.363 ms', 'D': '17.435 ms', 'E': '9.099 ms', 'F': '9.420 ms'}
efficient -> 13.765 ms | 1097.356 ns/elts | 45.723 ns/(elts*experts)
| 256*7*7 | 24 | 128 | 1097ns | 46ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 117.857 Mo
torch.cuda.max_memory_reserved: 132.121 Mo
---------------
computeAll -> 3.574 ms | 284.878 ns/elts | 11.87 ns/(elts*experts)
| 256*7*7 | 24 | 128 | 285ns | 12ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 967.43 Mo
torch.cuda.max_memory_reserved: 1.216 Go
---------------

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=64, nbFeatures=128, nbOut=128 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 56.024 Mo
torch.cuda.max_memory_reserved: 77.595 Mo
---------------
A:6.19%, B:37.76%, C1:4.91%, C2:29.79%, D:6.56%, E:3.36%, F:3.56%, {'A': '41.445 ms', 'B': '252.954 ms', 'C1': '32.914 ms', 'C2': '199.601 ms', 'D': '43.955 ms', 'E': '22.535 ms', 'F': '23.830 ms'}
efficient -> 33.498 ms | 2670.427 ns/elts | 41.725 ns/(elts*experts)
| 256*7*7 | 64 | 128 | 2670ns | 42ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 141.11 Mo
torch.cuda.max_memory_reserved: 159.384 Mo
---------------
computeAll -> 9.471 ms | 755.06 ns/elts | 11.798 ns/(elts*experts)
| 256*7*7 | 64 | 128 | 755ns | 12ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 2.53 Go
torch.cuda.max_memory_reserved: 3.112 Go
---------------

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=128, nbFeatures=128, nbOut=128 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 95.625 Mo
torch.cuda.max_memory_reserved: 111.149 Mo
---------------
A:5.86%, B:41.93%, C1:4.60%, C2:27.83%, D:6.12%, E:3.14%, F:3.28%, {'A': '19.868 ms', 'B': '142.207 ms', 'C1': '15.605 ms', 'C2': '94.382 ms', 'D': '20.761 ms', 'E': '10.644 ms', 'F': '11.132 ms'}
efficient -> 67.827 ms | 5407.124 ns/elts | 42.243 ns/(elts*experts)
| 256*7*7 | 128 | 128 | 5407ns | 42ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 180.646 Mo
torch.cuda.max_memory_reserved: 192.938 Mo
---------------
computeAll -> 24.483 ms | 1951.733 ns/elts | 15.248 ns/(elts*experts)
| 256*7*7 | 128 | 128 | 1952ns | 15ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 5.035 Go
torch.cuda.max_memory_reserved: 6.149 Go
---------------

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=256, nbFeatures=128, nbOut=128 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 175.907 Mo
torch.cuda.max_memory_reserved: 186.647 Mo
---------------
A:5.96%, B:40.75%, C1:4.71%, C2:27.98%, D:6.20%, E:3.22%, F:3.41%, {'A': '38.863 ms', 'B': '265.676 ms', 'C1': '30.681 ms', 'C2': '182.425 ms', 'D': '40.413 ms', 'E': '20.999 ms', 'F': '22.238 ms'}
efficient -> 130.381 ms | 10393.916 ns/elts | 40.601 ns/(elts*experts)
| 256*7*7 | 256 | 128 | 10394ns | 41ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 260.899 Mo
torch.cuda.max_memory_reserved: 289.407 Mo
---------------
computeAll -> 56.494 ms | 4503.704 ns/elts | 17.593 ns/(elts*experts)
| 256*7*7 | 256 | 128 | 4504ns | 18ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 10.048 Go
torch.cuda.max_memory_reserved: 12.231 Go
---------------


######################################################################################


*** unsing: baseBatchSize=256, convShape=(16, 16), nbExperts=6, nbFeatures=128, nbOut=128 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 51.018 Mo
torch.cuda.max_memory_reserved: 60.817 Mo
---------------
A:2.48%, B:73.45%, C1:2.14%, C2:12.96%, D:2.64%, E:1.36%, F:1.42%, {'A': '3.771 ms', 'B': '111.634 ms', 'C1': '3.260 ms', 'C2': '19.699 ms', 'D': '4.011 ms', 'E': '2.060 ms', 'F': '2.164 ms'}
efficient -> 7.599 ms | 115.957 ns/elts | 19.326 ns/(elts*experts)
| 256*16*16 | 6 | 128 | 116ns | 19ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 535.822 Mo
torch.cuda.max_memory_reserved: 570.425 Mo
---------------
computeAll -> 1.178 ms | 17.976 ns/elts | 2.996 ns/(elts*experts)
| 256*16*16 | 6 | 128 | 18ns | 3ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 1.361 Go
torch.cuda.max_memory_reserved: 1.766 Go
---------------

*** unsing: baseBatchSize=256, convShape=(16, 16), nbExperts=24, nbFeatures=128, nbOut=128 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 69.938 Mo
torch.cuda.max_memory_reserved: 92.275 Mo
---------------
A:4.86%, B:50.56%, C1:3.89%, C2:23.94%, D:5.15%, E:2.61%, F:2.75%, {'A': '15.061 ms', 'B': '156.619 ms', 'C1': '12.038 ms', 'C2': '74.152 ms', 'D': '15.950 ms', 'E': '8.076 ms', 'F': '8.535 ms'}
efficient -> 15.490 ms | 236.357 ns/elts | 9.848 ns/(elts*experts)
| 256*16*16 | 24 | 128 | 236ns | 10ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 520.138 Mo
torch.cuda.max_memory_reserved: 593.494 Mo
---------------
computeAll -> 4.552 ms | 69.465 ns/elts | 2.894 ns/(elts*experts)
| 256*16*16 | 24 | 128 | 69ns | 3ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 4.953 Go
torch.cuda.max_memory_reserved: 5.876 Go
---------------

*** unsing: baseBatchSize=256, convShape=(16, 16), nbExperts=64, nbFeatures=128, nbOut=128 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 111.444 Mo
torch.cuda.max_memory_reserved: 125.829 Mo
---------------
A:5.64%, B:43.56%, C1:4.69%, C2:26.67%, D:6.11%, E:3.05%, F:3.21%, {'A': '39.014 ms', 'B': '301.072 ms', 'C1': '32.409 ms', 'C2': '184.319 ms', 'D': '42.208 ms', 'E': '21.057 ms', 'F': '22.206 ms'}
efficient -> 34.557 ms | 527.294 ns/elts | 8.239 ns/(elts*experts)
| 256*16*16 | 64 | 128 | 527ns | 8ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 568.229 Mo
torch.cuda.max_memory_reserved: 591.397 Mo
---------------
computeAll -> 10.150 ms | 154.873 ns/elts | 2.42 ns/(elts*experts)
| 256*16*16 | 64 | 128 | 155ns | 2ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 13.036 Go
torch.cuda.max_memory_reserved: 15.316 Go
---------------


######################################################################################



*** unsing: baseBatchSize=256, convShape=(16, 16), nbExperts=6, nbFeatures=256, nbOut=256 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 94.025 Mo
torch.cuda.max_memory_reserved: 104.858 Mo
---------------
A:1.27%, B:86.80%, C1:0.99%, C2:6.76%, D:1.22%, E:0.62%, F:0.66%, {'A': '4.034 ms', 'B': '276.673 ms', 'C1': '3.142 ms', 'C2': '21.551 ms', 'D': '3.903 ms', 'E': '1.979 ms', 'F': '2.101 ms'}
efficient -> 15.938 ms | 243.188 ns/elts | 40.531 ns/(elts*experts)
| 256*16*16 | 6 | 256 | 243ns | 41ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 1.062 Go
torch.cuda.max_memory_reserved: 1.288 Go
---------------
computeAll -> 1.406 ms | 21.449 ns/elts | 3.575 ns/(elts*experts)
| 256*16*16 | 6 | 256 | 21ns | 4ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 2.712 Go
torch.cuda.max_memory_reserved: 3.509 Go
---------------

*** unsing: baseBatchSize=256, convShape=(16, 16), nbExperts=24, nbFeatures=256, nbOut=256 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 141.303 Mo
torch.cuda.max_memory_reserved: 163.578 Mo
---------------
A:3.28%, B:64.95%, C1:2.67%, C2:17.39%, D:3.53%, E:1.82%, F:1.88%, {'A': '15.308 ms', 'B': '302.876 ms', 'C1': '12.467 ms', 'C2': '81.102 ms', 'D': '16.458 ms', 'E': '8.509 ms', 'F': '8.763 ms'}
efficient -> 23.315 ms | 355.764 ns/elts | 14.824 ns/(elts*experts)
| 256*16*16 | 24 | 256 | 356ns | 15ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 1.061 Go
torch.cuda.max_memory_reserved: 1.086 Go
---------------
computeAll -> 4.840 ms | 73.855 ns/elts | 3.077 ns/(elts*experts)
| 256*16*16 | 24 | 256 | 74ns | 3ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 9.906 Go
torch.cuda.max_memory_reserved: 11.752 Go
---------------

*** unsing: baseBatchSize=256, convShape=(16, 16), nbExperts=64, nbFeatures=256, nbOut=256 ****
memory status of initial datas
torch.cuda.max_memory_allocated: 245.826 Mo
torch.cuda.max_memory_reserved: 260.047 Mo
---------------
A:4.85%, B:51.33%, C1:3.93%, C2:23.18%, D:5.18%, E:2.61%, F:2.77%, {'A': '40.052 ms', 'B': '423.501 ms', 'C1': '32.458 ms', 'C2': '191.239 ms', 'D': '42.755 ms', 'E': '21.501 ms', 'F': '22.875 ms'}
efficient -> 41.253 ms | 629.472 ns/elts | 9.835 ns/(elts*experts)
| 256*16*16 | 64 | 256 | 629ns | 10ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 1.134 Go
torch.cuda.max_memory_reserved: 1.21 Go
---------------
computeAll -> 0.150 μs | 0.002 ns/elts | 0.0 ns/(elts*experts)
| 256*16*16 | 64 | 256 | 0ns | 0ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 258.255 Mo
torch.cuda.max_memory_reserved: 301.99 Mo
---------------

*** unsing: baseBatchSize=256, convShape=(16, 16), nbExperts=4096, nbFeatures=256, nbOut=256 ****
nb params: 2_152_726_528
memory status of initial datas
torch.cuda.max_memory_allocated: 10.836 Go
torch.cuda.max_memory_reserved: 10.849 Go
---------------
A:6.00%, B:50.65%, C1:3.67%, C2:23.19%, D:4.79%, E:2.56%, F:2.67%, {'A': '177.811 ms', 'B': '1.502 sec', 'C1': '108.765 ms', 'C2': '687.508 ms', 'D': '141.880 ms', 'E': '76.007 ms', 'F': '79.094 ms'}
efficient -> 2.965 sec | 45235.869 ns/elts | 11.044 ns/(elts*experts)
| 256*16*16 | 4096 | 256 | 45236ns | 11ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 11.714 Go
torch.cuda.max_memory_reserved: 11.746 Go
---------------
computeAll -> 2.600 μs | 0.04 ns/elts | 0.0 ns/(elts*experts)
| 256*16*16 | 4096 | 256 | 0ns | 0ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 10.836 Go
torch.cuda.max_memory_reserved: 10.853 Go
---------------

```


### using efficient computation
| batch size | nb experts | nbFeatures | nb parameters | time | time/elt | time/elt/expert | peak memory used | time ratio | mem ratio |
|---|---|---|---|---|---|---|---|---|---|
| 256\*7*7 | 6 | 64 | 198_528 | 3.713 ms | 296ns | 49ns | 64Mo | 4.35x slower | 2.16x less |
| 256\*7*7 | 24 | 64 | 794_112 | 11.261 ms | 898ns | 37ns | 62Mo | 3.62x slower | 7.84x less |
| 256\*7*7 | 64 | 64 | 2_117_632 | 34.244 ms | 2'730ns | 43ns | 71Mo | 3.30x slower | 17.82x less |
| 256\*7*7 | 128 | 64 | 4_235_264 | 70.829 ms | 5'646ns | 44ns | 86Mo | 2.36x slower | 29.22x less |
| 256\*7*7 | 256 | 64 | 8_470_528 | 118.720 ms | 9'464ns | 37ns | 115Mo | 1.30x slower | 43.55x less |
| 256\*7*7 | 512 | 64 | 16_941_056 | 257.920 ms | 20'561ns | 40ns | 174Mo | 1.18x slower | 57.55x less |
|---|---|---|---|---|---|---|---|---|
| 256\*7*7 | 6 | 128 | 790_272 | 4.172 ms | 333ns | 55ns | 112Mo | 3.78x slower | 2.41x less |
| 256\*7*7 | 24 | 128 | 3_161_088 | 13.765 ms | 1'097ns | 46ns | 117Mo | 3.85x slower | 8.26x less |
| 256\*7*7 | 64 | 128 | 8_429_568 | 33.498 ms | 2'670ns | 42ns | 141Mo | 3.54x slower | 17.94x less |
| 256\*7*7 | 128 | 128 | 16_859_136 | 67.827 ms | 5'407ns | 42ns | 180Mo | 2.77x slower | 27.97x less |
| 256\*7*7 | 256 | 128 | 33_718_272 | 130.381 ms | 1'0394ns | 41ns | 260Mo | 2.31x slower | 38.65x less | 
|---|---|---|---|---|---|---|---|---|
| 256\*16*16 | 6 | 128 | 790_272 | 7.599 ms | 122ns | 20ns | 534Mo | 6.78x slower | 2.55x less |
| 256\*16*16 | 24 | 128 | 3_161_088 | 15.490 ms | 251ns | 10ns | 520Mo | 3.64x slower | 9.53x less |
| 256\*16*16 | 64 | 128 | 8_429_568 | 34.557 ms | 527ns | 8ns | 568Mo | 3.40x slower | 22.95x less |
|---|---|---|---|---|---|---|---|---|
| 256\*16*16 | 6 | 256 | 3_153_408 | 15.938 ms | 243ns | 41ns | 1'062Mo | 11.57x slower | 2.55x less |
| 256\*16*16 | 24 | 256 | 12_613_632 | 23.315 ms | 356ns | 15ns | 1'061Mo | 4.81x slower | 9.34x less |
| 256\*16*16 | 64 | 256 | 33_636_352 | 41.253 ms | 602ns | 9.4ns | 1'136Mo | / | ~ 22x less |
| 256\*16*16 | 4096 | 256 | 2_152_726_528 | 2'965.0 ms | 45'236ns | 11ns | 11'714Mo | / | ~ 145x less |

### using computation with all experts
| batch size | nb experts | nbFeatures | time | time/elt | time/elt/expert | peak memory used |
|---|---|---|---|---|---|---|
| 256\*7*7 | 6 | 64 | 0.855 ms | 68ns | 11ns | 138Mo |
| 256\*7*7 | 24 | 64 | 3.107 ms | 248ns | 10ns | 486Mo |
| 256\*7*7 | 64 | 64 | 10.371 ms | 827ns | 13ns | 1'265Mo |
| 256\*7*7 | 128 | 64 | 30.070 ms | 2397ns | 19ns | 2'513Mo |
| 256\*7*7 | 256 | 64 | 91.033 ms | 7257ns | 28ns | 5'008Mo |
| 256\*7*7 | 512 | 64 | 217.858 ms | 17368ns | 34ns | 10'013Mo |
|---|---|---|---|---|---|---|
| 256\*7*7 | 6 | 128 | 1.099 ms | 88ns | 15ns | 270Mo |
| 256\*7*7 | 24 | 128 | 3.574 ms | 285ns | 12ns | 967Mo |
| 256\*7*7 | 64 | 128 | 9.471 ms | 755ns | 12ns | 2'530Mo |
| 256\*7*7 | 128 | 128 | 24.483 ms | 1952ns | 15ns | 5'035Mo |
| 256\*7*7 | 256 | 128 | 56.494 ms | 4504ns | 18ns | 10'048Mo |
|---|---|---|---|---|---|---|
| 256\*16*16 | 6 | 128 | 1.178 ms | 18ns | 3.0ns | 1'361Mo |
| 256\*16*16 | 24 | 128 | 4.552 ms | 69ns | 2.9ns | 4'953Mo |
| 256\*16*16 | 64 | 128 | 10.150 ms | 155ns | 2.4ns | 13'036Mo |
|---|---|---|---|---|---|---|
| 256\*16*16 | 6 | 256 | 1.406 ms | 21ns | 3.5ns | 2'712Mo |
| 256\*16*16 | 24 | 256 | 4.840 ms | 74ns | 3.1ns | 9'906Mo |
| 256\*16*16 | 64 | 128 | / | / | / | ~ 26Go |
| 256\*16*16 | 4096 | 256 | / | / | / | ~ 1.7To |

### evolution of times per steps for the hard routing (GPU)
| batch size | nb experts | nb features | nb params | total time | time without 'where' | 'where' time | 'expert' time | time 'other' |
|---|---|---|---|---|---|---|---|---|
| 256\*7*7 | 6 | 64 | 198_528 | 4.729ms | 2.219ms | 2.511ms | 1.162ms | 1.056ms |
| 256\*7*7 | 6 | 128 | 790_272 | 6.478ms | 2.873ms | 3.605ms | 1.520ms | 1.353ms |
| 256\*7*7 | 6 | 256 | 3_153_408 | 9.606ms | 3.652ms | 5.954ms | 2.398ms | 1.254ms |
| 256\*7*7 | 6 | 512 | 12_598_272 | 17.885ms | 4.477ms | 13.407ms | 2.930ms | 1.548ms |
| 256\*7*7 | 6 | 1024 | 50_362_368 | 38.086ms | 5.219ms | 32.867ms | 3.731ms | 1.488ms |


### evolution of times per steps for the hard routing (CPU)
| batch size | nb experts | nb features | nb params | total time | time without 'where' | 'where' time | 'expert' time | time 'other' |
|---|---|---|---|---|---|---|---|---|
| 256\*7*7 | 6 | 64 | 198_528 | 13.522ms | 12.951ms | 0.571ms | 8.889ms | 4.061ms |
| 256\*7*7 | 6 | 128 | 790_272 | 34.423ms | 33.850ms | 0.573ms | 29.465ms | 4.385ms |
| 256\*7*7 | 6 | 256 | 3_153_408 | 87.356ms | 86.722ms | 0.634ms | 78.544ms | 8.179ms |
| 256\*7*7 | 6 | 512 | 12_598_272 | 237.992ms | 237.292ms | 0.700ms | 221.292ms | 16.000ms |
| 256\*7*7 | 6 | 1024 | 50_362_368 | 800.219ms | 799.553ms | 0.666ms | 770.937ms | 28.616ms |


### evolution of times per steps for the hard routing (GPU)
| batch size | nb experts | nb features | nb params | total time | avg/expert | 'where' avg/expert | 'expert' avg/expert | 'other' avg/expert |
|---|---|---|---|---|---|---|---|---|
| 256\*7*7 | 6 | 128 | 790_272 | 6.478ms | 1.080ms | 0.601ms | 0.253ms | 0.226ms |
| 256\*7*7 | 12 | 128 | 1_580_544 | 11.097ms | 0.925ms | 0.415ms | 0.268ms | 0.242ms |
| 256\*7*7 | 24 | 128 | 3_161_088 | 15.769ms | 0.657ms | 0.262ms | 0.219ms | 0.176ms |
| 256\*7*7 | 48 | 128 | 6_322_176 | 28.468ms | 0.593ms | 0.209ms | 0.204ms | 0.181ms |
| 256\*7*7 | 96 | 128 | 12_644_352 | 55.015ms | 0.573ms | 0.193ms | 0.199ms | 0.181ms |


### evolution of times per steps for the hard routing (CPU)
| batch size | nb experts | nb features | nb params | total time | avg/expert | 'where' avg/expert | 'expert' avg/expert | 'other' avg/expert |
|---|---|---|---|---|---|---|---|---|
| 256\*7*7 | 6 | 128 | 790_272 | 31.891ms | 5.315ms | 0.098ms | 4.428ms | 0.790ms |
| 256\*7*7 | 12 | 128 | 1_580_544 | 35.967ms | 2.997ms | 0.065ms | 2.357ms | 0.575ms |
| 256\*7*7 | 24 | 128 | 3_161_088 | 44.140ms | 1.839ms | 0.048ms | 1.177ms | 0.614ms |
| 256\*7*7 | 48 | 128 | 6_322_176 | 50.881ms | 1.060ms | 0.042ms | 0.661ms | 0.357ms |
| 256\*7*7 | 96 | 128 | 12_644_352 | 69.499ms | 0.724ms | 0.036ms | 0.391ms | 0.297ms |

```raw
typical runs for the efficient mode:

CPU:

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=24, nbFeatures=128, nbOut=128 ****
nb params: 3_161_088 (nbHidden: 512)
topK:76.47%, softMax:14.29%, expertsWeigths:9.25%, {'topK': '0.924 ms', 'softMax': '0.173 ms', 'expertsWeigths': '0.112 ms'}
memory status of initial datas
torch.cuda.max_memory_allocated: 34.41 Mo
torch.cuda.max_memory_reserved: 67.109 Mo
---------------
A:9.40%, B:3.77%, C1:5.64%, C2:64.77%, D:3.70%, E:2.97%, F:6.45%, {'A': '33.777 ms', 'B': '13.556 ms', 'C1': '20.273 ms', 'C2': '232.732 ms', 'D': '13.285 ms', 'E': '10.681 ms', 'F': '23.165 ms'}
efficient -> 35.932 ms | 2864.48 ns/elts | 119.353 ns/(elts*experts)
| 256*7*7 | 24 | 128 | 2864ns | 119ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 34.41 Mo
torch.cuda.max_memory_reserved: 67.109 Mo
---------------
computeAll -> 413.034 ms | 32926.852 ns/elts | 1371.952 ns/(elts*experts)
| 256*7*7 | 24 | 128 | 32927ns | 1372ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 34.41 Mo
torch.cuda.max_memory_reserved: 67.109 Mo


GPU:

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=24, nbFeatures=128, nbOut=128 ****
nb params: 3_161_088 (nbHidden: 512)
topK:41.80%, softMax:16.20%, expertsWeigths:41.99%, {'topK': '43.600 μs', 'softMax': '16.900 μs', 'expertsWeigths': '43.800 μs'}
memory status of initial datas
torch.cuda.max_memory_allocated: 56.287 Mo
torch.cuda.max_memory_reserved: 71.303 Mo
---------------
A:5.08%, B:46.87%, C1:4.02%, C2:26.76%, D:5.23%, E:2.75%, F:2.85%, {'A': '7.387 ms', 'B': '68.219 ms', 'C1': '5.847 ms', 'C2': '38.944 ms', 'D': '7.619 ms', 'E': '3.999 ms', 'F': '4.143 ms'}
efficient -> 14.553 ms | 1160.184 ns/elts | 48.341 ns/(elts*experts)
| 256*7*7 | 24 | 128 | 1160ns | 48ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 143.222 Mo
torch.cuda.max_memory_reserved: 167.772 Mo
---------------
computeAll -> 4.275 ms | 340.789 ns/elts | 14.2 ns/(elts*experts)
| 256*7*7 | 24 | 128 | 341ns | 14ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 992.883 Mo
torch.cuda.max_memory_reserved: 1.223 Go
---------------
```


| 256*7*7 | 6 | 128 | 2389ns | 398ns | Mo |
| 256*7*7 | 12 | 128 | 2569ns | 214ns | Mo |
| 256*7*7 | 24 | 128 | 3098ns | 129ns | Mo |

| 256*7*7 | 6 | 128 | 6808ns | 1135ns | Mo |
| 256*7*7 | 12 | 128 | 13712ns | 1143ns | Mo |
| 256*7*7 | 24 | 128 | 32927ns | 1372ns | Mo |





| 256*7*7 | 6 | 128 | 322ns | 54ns | Mo |
| 256*7*7 | 12 | 128 | 543ns | 45ns | Mo |
| 256*7*7 | 24 | 128 | 1021ns | 43ns | Mo |

| 256*7*7 | 6 | 128 | 125ns | 21ns | Mo |
| 256*7*7 | 12 | 128 | 171ns | 14ns | Mo |
| 256*7*7 | 24 | 128 | 391ns | 16ns | Mo |


In [308]:
class GatingWithNoise(torch.nn.Module):
    def __init__(self, nbIn:int, nbExperts:int) -> None:
        super().__init__()
        self.gate = torch.nn.Linear(nbIn, nbExperts)
        self.noise = torch.nn.Linear(nbIn, nbExperts)
        
    def __call__(self, x:torch.Tensor)->torch.Tensor:
        return super().__call__(x)
    
    def forward(self, x:torch.Tensor)->torch.Tensor:
        """x: (batch, features)
        return (batch, experts)"""
        gateOut = self.gate(x)
        noiseOut = torch.nn.functional.softplus(self.noise(x))
        trueNoise = torch.randn_like(noiseOut)
        return gateOut + (trueNoise * noiseOut)
        
class Block_FFD_Moe(torch.nn.Module):
    def __init__(
            self, nbIn:int, nbOut:int, nbExperts:int,
            topK:int, fromCNN:bool, memMode:bool=False) -> None:
        super().__init__()
        self.topK: int = topK
        self.fromCNN: bool = fromCNN
        self.memEfficientMode: bool = memMode
        self.nbOut: int = nbOut
        self.gating = GatingWithNoise(nbIn, nbExperts)
        nbHidden = max(nbIn, nbOut) * 4
        self.experts = torch.nn.ModuleList([
            torch.nn.Sequential(
                torch.nn.Linear(nbIn, nbHidden), torch.nn.ReLU(),
                torch.nn.Linear(nbHidden, nbOut))
            for _ in range(nbExperts)])

    @property
    def nbExperts(self)->int:
        return len(self.experts)

    def __call__(self, x:torch.Tensor)->tuple[torch.Tensor, torch.Tensor]:
        return super().__call__(x)
    
    def forward(self, x:torch.Tensor)->tuple[torch.Tensor, torch.Tensor]:
        """x: [normal](batch, features) | [CNN](batch, channel, H, W)"""
        if self.fromCNN:
            batchSize, nbFeatures, H, W = x.shape
            x = x.permute((0, 2, 3, 1)) # (B, H, W, C)
            x = x.reshape((batchSize*H*W), nbFeatures) # (B * H * W, C)
        else:
            H, W = (0, 0)
            batchSize, nbFeatures = x.shape
        # => x: (batch, features)
        batchSize2, nbFeatures = x.shape
        
        gatingLogits = self.gating(x) # (batch, experts)
        gatingLogits_k, topK_indices = torch.topk(gatingLogits, k=self.topK, dim=1)
        # (batch, k), (batch, k)
        gatingProb = torch.softmax(gatingLogits_k, dim=1) # (batch, k)
        allExperts_weigths = torch.zeros((batchSize2, self.nbExperts), device=x.device) # (batch, experts)
        allExperts_weigths.scatter_(dim=1, index=topK_indices, src=gatingProb)

        if self.memEfficientMode is True:
            xOut = torch.zeros((batchSize2, self.nbOut), device=x.device)
            for iExpert in range(self.nbExperts):
                expertUse, *_ = torch.where((topK_indices == iExpert).any(dim=1))
                if expertUse.shape[0] == 0:
                    continue # expert isn't used
                expertOuts = self.experts[iExpert](x[expertUse])
                expertGates = allExperts_weigths[expertUse, iExpert].unsqueeze(dim=1)
                res = (expertGates * expertOuts)
                xOut.index_add_(dim=0, index=expertUse, source=res)
        else: 
            expertsOut = torch.stack([expert(x) for expert in self.experts], dim=2)
            xOut = torch.sum(allExperts_weigths[:, None, :] * expertsOut, dim=2)
        
        if self.fromCNN:
            xOut = xOut.reshape(batchSize, H, W, self.nbOut) # (B, H, W, C)
            xOut = xOut.permute(0, 3, 1, 2) # (B, C, H, W)
            allExperts_weigths = allExperts_weigths.reshape(batchSize, H, W, self.nbExperts) # (B, H, W, experts)
        return (xOut, allExperts_weigths)

In [309]:
def findHiddenForEqParams(targetParams:int, nbFeatures:int, nbOut:int)->int:
    nbHidden = (targetParams - nbOut)/(nbFeatures + nbOut + 1)
    return nbHidden.__ceil__()

In [311]:
device = torch.device("cuda"); print(device.type)

def run_compare_eff_fast_eqParams(
        baseBatchSize:int, convShape:tuple[int, int], 
        nbExperts:int, topK:int, nbFeatures:int, nbOut:int):
    print(f"*** unsing: {baseBatchSize=}, {convShape=}, {nbExperts=}, {nbFeatures=}, {nbOut=} ****")
    resetMemStats()
    batchSize = baseBatchSize * convShape[0] * convShape[1]
    
    moeEfficient = Block_FFD_Moe(nbIn=nbFeatures, nbOut=nbOut, nbExperts=nbExperts, topK=topK, fromCNN=False, memMode=True).to(device)
    moeFast = Block_FFD_Moe(nbIn=nbFeatures, nbOut=nbOut, nbExperts=nbExperts, topK=topK, fromCNN=False, memMode=False).to(device)
    paramsEff = sum(product(p.shape) for p in moeEfficient.parameters())
    paramsFast = sum(product(p.shape) for p in moeFast.parameters())
    assert paramsEff == paramsFast
    nbHiddenEqParams = findHiddenForEqParams(paramsEff, nbFeatures=nbFeatures, nbOut=nbOut)
    ffdEqParams = torch.nn.Sequential(
        torch.nn.Linear(nbFeatures, nbHiddenEqParams),
        torch.nn.ReLU(), torch.nn.Linear(nbHiddenEqParams, nbOut)).to(device)
    print(f"nb params moe eff: {paramsEff:_d}")
    print(f"nb params moe fast: {paramsFast:_d}")
    print(f"nb params equiv params: {sum(product(p.shape) for p in ffdEqParams.parameters()):_d}")
    
    datasIn = torch.randn((batchSize, nbFeatures), device=device)
    memDiag("initial datas")

    N = 1
    with SimpleProfiler("efficient") as sp1:
        for i in range(N):
            _ = moeEfficient(datasIn)
            del _
    mem1 = memDiag("after effiecient topK")
    showTimeStat(sp1, nbIters=N, baseBatchSize=baseBatchSize, convShape=convShape, nbExperts=nbExperts, nbFeatures=nbFeatures, memUsage=mem1)

    with SimpleProfiler("computeAll") as sp2:
        for i in range(N):
            _ = moeFast(datasIn)
            del _
    mem2 = memDiag("after compute all")
    showTimeStat(sp2, nbIters=N, baseBatchSize=baseBatchSize, convShape=convShape, nbExperts=nbExperts, nbFeatures=nbFeatures, memUsage=mem2)
    
    with SimpleProfiler("equivalent Params") as sp3:
        for i in range(N):
            _ = ffdEqParams(datasIn)
            del _
    mem3 = memDiag("after equivalent params")
    showTimeStat(sp3, nbIters=N, baseBatchSize=baseBatchSize, convShape=convShape, nbExperts=1, nbFeatures=nbFeatures, memUsage=mem3)
    
run_compare_eff_fast_eqParams(baseBatchSize=256, convShape=(7, 7), nbExperts=6, topK=2, nbFeatures=2048, nbOut=2048)

cuda
*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=6, nbFeatures=2048, nbOut=2048 ****
nb params moe eff: 201_412_620
nb params moe fast: 201_412_620
nb params equiv params: 201_414_665
memory status of initial datas
torch.cuda.max_memory_allocated: 2.539 Go
torch.cuda.max_memory_reserved: 2.554 Go
---------------
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 4.153 Go
torch.cuda.max_memory_reserved: 4.343 Go
---------------
efficient -> 98.752 ms | 7872.425 ns/elts | 1312.071 ns/(elts*experts)
| 256\*7*7 | 6 | 2048 | 98.752 ms | 7872ns | 1312ns | 4_152Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 6.343 Go
torch.cuda.max_memory_reserved: 7.487 Go
---------------
computeAll -> 20.158 ms | 1606.999 ns/elts | 267.833 ns/(elts*experts)
| 256\*7*7 | 6 | 2048 | 20.158 ms | 1607ns | 268ns | 6_342Mo |
memory status of after equivalent params
torch.cuda.max_memory_allocated: 7.472 Go
torch.cuda.max_memory_reserved: 7.491 Go
------

In [99]:
device = torch.device("cuda"); print(device.type)

def run_calcFLOPs_eff_fast_eqParams(
        baseBatchSize:int, convShape:tuple[int, int], 
        nbExperts:int, topK:int, nbFeatures:int, nbOut:int):
    print(f"*** unsing: {baseBatchSize=}, {convShape=}, {nbExperts=}, {nbFeatures=}, {nbOut=} ****")
    resetMemStats()
    batchSize = baseBatchSize * convShape[0] * convShape[1]
    
    moeEfficient = Block_FFD_Moe(nbIn=nbFeatures, nbOut=nbOut, nbExperts=nbExperts, topK=topK, fromCNN=False, memMode=True).to(device)
    moeFast = Block_FFD_Moe(nbIn=nbFeatures, nbOut=nbOut, nbExperts=nbExperts, topK=topK, fromCNN=False, memMode=False).to(device)
    paramsEff = sum(product(p.shape) for p in moeEfficient.parameters())
    paramsFast = sum(product(p.shape) for p in moeFast.parameters())
    assert paramsEff == paramsFast
    nbHiddenEqParams = findHiddenForEqParams(paramsEff, nbFeatures=nbFeatures, nbOut=nbOut)
    ffdEqParams = torch.nn.Sequential(
        torch.nn.Linear(nbFeatures, nbHiddenEqParams),
        torch.nn.ReLU(), torch.nn.Linear(nbHiddenEqParams, nbOut)).to(device)
    print(f"nb params moe eff: {paramsEff:_d}")
    print(f"nb params moe fast: {paramsFast:_d}")
    print(f"nb params equiv params: {sum(product(p.shape) for p in ffdEqParams.parameters()):_d}")
    
    datasIn = torch.randn((batchSize, nbFeatures), device=device)
    flopsEff, macsEff, _ = calflops.calculate_flops(
        model=moeEfficient, input_shape=tuple(datasIn.shape), output_as_string=False, output_precision=2, print_results=False)
    print(f"moe eff: {float(flopsEff)/1e9:.2f} GFLOPs")
    flopsFast, macsFast, _ = calflops.calculate_flops(
        model=moeFast, input_shape=tuple(datasIn.shape), output_as_string=False, output_precision=2, print_results=False)
    print(f"moe fast: {float(flopsFast)/1e9:.2f} GFLOPs")
    flopsEqParams, macsEqParams, _ = calflops.calculate_flops(
        model=ffdEqParams, input_shape=tuple(datasIn.shape), output_as_string=False, output_precision=2, print_results=False)
    print(f"eq params: {float(flopsEqParams)/1e9:.2f} GFLOPs")
    
    
run_calcFLOPs_eff_fast_eqParams(baseBatchSize=256, convShape=(7, 7), nbExperts=6, topK=2, nbFeatures=512, nbOut=512)

cuda
*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=6, nbFeatures=512, nbOut=512 ****
nb params moe eff: 12_604_428
nb params moe fast: 12_604_428
nb params equiv params: 12_604_937
moe eff: 527.16 GFLOPs
moe fast: 1579.94 GFLOPs
eq params: 1580.33 GFLOPs


```raw

experiments results: 
compare full MOE with different modes


cpu
*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=6, nbFeatures=128, nbOut=128 ****
nb params moe eff: 791_820
nb params moe fast: 791_820
nb params equiv params: 791_945
memory status of initial datas
torch.cuda.max_memory_allocated: 8.52 Mo
torch.cuda.max_memory_reserved: 20.972 Mo
---------------
efficient -> 30.221 ms | 2409.232 ns/elts | 401.539 ns/(elts*experts)
| 256*7*7 | 6 | 128 | 2409ns | 402ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 8.52 Mo
torch.cuda.max_memory_reserved: 20.972 Mo
---------------
computeAll -> 85.408 ms | 6808.65 ns/elts | 1134.775 ns/(elts*experts)
| 256*7*7 | 6 | 128 | 6809ns | 1135ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 8.52 Mo
torch.cuda.max_memory_reserved: 20.972 Mo
---------------
equivalent Params -> 72.083 ms | 5746.381 ns/elts | 5746.381 ns/(elts*experts)
| 256*7*7 | 1 | 128 | 5746ns | 5746ns | Mo |
memory status of after equivalent params
torch.cuda.max_memory_allocated: 8.52 Mo
torch.cuda.max_memory_reserved: 20.972 Mo
---------------


cpu
*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=64, nbFeatures=128, nbOut=128 ****
nb params moe eff: 8_446_080
nb params moe fast: 8_446_080
nb params equiv params: 8_446_176
memory status of initial datas
torch.cuda.max_memory_allocated: 8.52 Mo
torch.cuda.max_memory_reserved: 20.972 Mo
---------------
efficient -> 69.890 ms | 5571.612 ns/elts | 87.056 ns/(elts*experts)
| 256*7*7 | 64 | 128 | 5572ns | 87ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 8.52 Mo
torch.cuda.max_memory_reserved: 20.972 Mo
---------------
computeAll -> 1.336 sec | 106507.039 ns/elts | 1664.172 ns/(elts*experts)
| 256*7*7 | 64 | 128 | 106507ns | 1664ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 8.52 Mo
torch.cuda.max_memory_reserved: 20.972 Mo
---------------
equivalent Params -> 1.541 sec | 122837.014 ns/elts | 122837.014 ns/(elts*experts)
| 256*7*7 | 1 | 128 | 122837ns | 122837ns | Mo |
memory status of after equivalent params
torch.cuda.max_memory_allocated: 8.52 Mo
torch.cuda.max_memory_reserved: 20.972 Mo
---------------




cuda
*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=6, nbFeatures=128, nbOut=128 ****
nb params moe eff: 791_820
nb params moe fast: 791_820
nb params equiv params: 791_945
memory status of initial datas
torch.cuda.max_memory_allocated: 24.447 Mo
torch.cuda.max_memory_reserved: 29.36 Mo
---------------
efficient -> 7.419 ms | 591.406 ns/elts | 98.568 ns/(elts*experts)
| 256*7*7 | 6 | 128 | 591ns | 99ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 128.635 Mo
torch.cuda.max_memory_reserved: 157.286 Mo
---------------
computeAll -> 3.468 ms | 276.499 ns/elts | 46.083 ns/(elts*experts)
| 256*7*7 | 6 | 128 | 276ns | 46ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 263.686 Mo
torch.cuda.max_memory_reserved: 356.516 Mo
---------------
equivalent Params -> 1.172 ms | 93.399 ns/elts | 93.399 ns/(elts*experts)
| 256*7*7 | 1 | 128 | 93ns | 93ns | Mo |
memory status of after equivalent params
torch.cuda.max_memory_allocated: 334.825 Mo
torch.cuda.max_memory_reserved: 339.739 Mo
---------------


cuda
*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=64, nbFeatures=128, nbOut=128 ****
nb params moe eff: 8_446_080
nb params moe fast: 8_446_080
nb params equiv params: 8_446_176
memory status of initial datas
torch.cuda.max_memory_allocated: 116.297 Mo
torch.cuda.max_memory_reserved: 127.926 Mo
---------------
efficient -> 41.922 ms | 3342.02 ns/elts | 52.219 ns/(elts*experts)
| 256*7*7 | 64 | 128 | 3342ns | 52ns | Mo |
memory status of after effiecient topK
torch.cuda.max_memory_allocated: 227.938 Mo
torch.cuda.max_memory_reserved: 262.144 Mo
---------------
computeAll -> 38.071 ms | 3034.997 ns/elts | 47.422 ns/(elts*experts)
| 256*7*7 | 64 | 128 | 3035ns | 47ns | Mo |
memory status of after compute all
torch.cuda.max_memory_allocated: 2.602 Go
torch.cuda.max_memory_reserved: 3.181 Go
---------------
equivalent Params -> 8.122 ms | 647.449 ns/elts | 647.449 ns/(elts*experts)
| 256*7*7 | 1 | 128 | 647ns | 647ns | Mo |
memory status of after equivalent params
torch.cuda.max_memory_allocated: 3.414 Go
torch.cuda.max_memory_reserved: 3.429 Go
---------------


*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=6, nbFeatures=128, nbOut=128 ****
nb params moe eff: 791_820
nb params moe fast: 791_820
nb params equiv params: 791_945
moe eff: 13.26 GFLOPs
moe fast: 39.61 GFLOPs
eq params: 39.65 GFLOPs

*** unsing: baseBatchSize=256, convShape=(7, 7), nbExperts=64, nbFeatures=128, nbOut=128 ****
nb params moe eff: 8_446_080
nb params moe fast: 8_446_080
nb params equiv params: 8_446_176
moe eff: 14.00 GFLOPs
moe fast: 422.55 GFLOPs
eq params: 422.96 GFLOPs

```

### comparaison of behaviours on different devices

| mode | batch size | nb experts | nbFeatures | nb parameters | CPU time | GPU time | peak GPU memory | FLOPs |
|---|---|---|---|---|---|---|---|---|
| hard routage | 256\*7*7 | 6 | 128 | 791'820 | 30.2 ms | 7.42 ms | 129 Mo | 13.3 GFLOPs |
| soft routage | 256\*7*7 | 6 | 128 | 791'820 | 85.4 ms | 3.47 ms | 264 Mo | 39.6 GFLOPs |
| simple FFD | 256\*7*7 | 6 | 128 | 791'945 | 72.0 ms | 1.17 ms | 335 Mo | 39.7 GFLOPs |
|---|---|---|---|---|---|---|---|
| hard routage | 256\*7*7 | 64 | 128 | 8'446'080 | 69.8 ms | 41.92 ms | 228 Mo | 14.0 GFLOPs |
| soft routage | 256\*7*7 | 64 | 128 | 8'446'080 | 1'336 ms | 38.07 ms | 2'602 Mo | 422.6 GFLOPs |
| simple FFD | 256\*7*7 | 64 | 128 | 8'446'176 | 1'541 ms | 8.12 ms | 3'414 Mo | 423.0 GFLOPs |